src/dataviews/dataview-model-base.js

```diff
@@ -76,6 +78,7 @@ module.exports = Model.extend({
 
   _initBinds: function () {
     this.listenTo(this.layer, 'change:visible', this._onLayerVisibilityChanged);
+    this.listenTo(this.layer, 'change:source', this._onLayerVisibilityChanged);
```

Comment:
Yes, it's my understanding by looking at the code and take into consideration that the source will change from time to time (e.g. adding/remove nodes in the builder). 
So when the head node of the layer (i.e. source) changes we would need to make sure [the correct analysis](https://github.com/CartoDB/cartodb.js/blame/1357-fix-dataviews-analysis-states/src/dataviews/dataview-model-base.js#L103) is listened to.